pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java

```diff
@@ -278,5 +287,37 @@ private boolean isPartitionMatch(FilterQueryTree filterQueryTree, PartitionInfo
       _partitionFunction = partitionFunction;
       _partitions = partitions;
     }
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      PartitionInfo that = (PartitionInfo) o;
+      return Objects.equals(_partitionFunction, that._partitionFunction) &&
+        Objects.equals(_partitions, that._partitions);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(_partitionFunction, _partitions);
+    }
+  }
+
+  /**
+   * This class is created to speed up partition number calls. Sometimes with lots of segments (like 40k+)
+   * Computing partition values may be CPU intensive, caching the computation result may help
+   */
+  private static class CachedPartitionFunction {
+    final Map<PartitionFunction, Map<String, Integer>> cache = new HashMap<>();
+
+    public int getPartition(PartitionFunction func, String value) {
```

Comment:
Can you elaborate a bit more on the meaning of the `value` here? Does that mean all the values will be cached in the hashmap and will never get cleaned up?

Reply:
For a partition function, it should be a stable fixed map: for each specific function f(x) and a string input x, we should always get the same partition for that immutable string x.
The only major concern could be the cached map size; if we have too many values in the query we could end up having OOM case.

Let me change to use Guava's CacheBuilder if possible https://guava.dev/releases/17.0/api/docs/com/google/common/cache/CacheBuilder.html